osquery/process/processes_stats.cpp

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include "processes_stats.h"
+
+#include <iostream>
+#include <sstream>
+
+#include <osquery/logger/logger.h>
+#include <osquery/utils/system/time.h>
+
+namespace osquery {
+
+CLI_FLAG(bool,
+         enable_processes_stats,
+         false,
+         "Enables the service that collects processes stats");
+
+CLI_FLAG(uint32,
+         processes_stats_interval,
+         2,
+         "Seconds that the processes stats service waits between scans");
+
+CLI_FLAG(
+    bool,
+    processes_stats_worker_only,
+    false,
+    "Enables process stats collection only for the osquery worker process");
+
+CLI_FLAG(uint32,
+         processes_stats_cpu_peak_samples,
+         30,
+         "How many past cpu usage samples to use to calculate the cpu peak "
+         "usage");
```

Comment:
This is going to be confusing to use. It requires the user to understand that this means it's the peak value within `processes_stats_cpu_peak_samples * processes_stats_interval`.

I don't know if it makes sense, but it might be more approachable to maintain the traditional 1min, 5min, 15min values.

Reply:
Or maybe this is the simplest pattern

Reply:
Well the thing is that the expiration and the math is done at the sample boundary, not minutes or seconds.

If you have an interval of 5 seconds for the samples, you have the samples for the time at 0, 5, 10, 15, 20, 25, 30, 35 seconds.
If you say that you want calculate the peak usage over the last 28 seconds, that won't be the reality, it will be an average of samples between 0 and 25 and there won't be a sample for second 26, 27, 28.. or any nanosecond in between.

So the end user will still have to correlate what those `last 28 seconds` really mean with a specific sample interval.
This was my reasoning, but if you feel it's clearer to provide that range in seconds, then I can change the logic.

Reply:
It seems like a fine thing to try - I suspect most people won't change these values.

I agree, that while seconds are more natural, it gets awkward since the boundaries line up.

I think reasoning about the 1,5,15 approach makes sense, but supporting 15 is a bunch more data

Reply:
Sorry I'm not sure I follow the comment here. There's only a single window of samples being kept to calculate the peak.

Were you suggesting the `processes_stats_cpu_peak_samples` flags to have fixed choices, 1 or 5 or 15min, instead of a free choice in samples (or seconds?).

And are you also saying that 15min seem to much samples to keep alive?
If so, we know of interest in keeping that data for longer, like 20 minutes or so. Also when you're only keeping it for the worker, the memory overhead is negligible. We are talking about 2 bytes for each sample, over 20mins at a sample rate of one sample every 5 seconds, it's 240 samples = 480 bytes.

Reply:
@directionless do you want to assume @Smjert  is correct that sampling CPU usage over time results in negligible data, or do you want to see an example? 

@Smjert I would tend to agree with the request to expose fewer control variables to the user. If measuring CPU usage is a matter of polling/sampling at a `sampling_interval` or `sampling_frequency` then we might just choose a sane default for that. The user is more likely to just want to always see the three columns like `average_over_the_last_1_min`, `average_over_the_last_5_min`, `average_over_the_last_15_min`

Reply:
Well the problem is that measuring cpu_usage and the peaks totally depends on the information you want to get out and the performance of doing so, which depends on the number of processes running.
I don't think it's good to force some sampling interval because it's what we now think is correct.

So here there are two interests colliding maybe: 

1. There's a general interest of having osquery track the cpu usage of a set of processes running in the system. This can have very different requests from the user, depending on what they are interested in.
    - Are they interested in seeing why a certain process sometimes have very high and quick CPU spikes?
       Then they will need a very frequent sampling interval to get the spike properly, otherwise it gets averaged out.
    - Or are they interested in averaging out the usage so that in general they want to understand how active is a process?
       Then they could use longer sampling intervals.
    - At the same time, what the process count?
      They might be forced to increase the sampling interval if it's too heavy.
    - How often are they able/want to query?
      Choosing how many samples to keep for the last peak means you have chances to see "all" the data you're interested in, without losing some.
2. They want to track osquery own CPU usage only, to understand when it's acting up and when the watchdog reacts. The problem with this is that the watchdog uses its own sampling interval, and so if you don't use something very similar (even if the samples will always be desynced), then it's basically impossible to compare usages.
Moreover the watchdog reacts on a certain amount of CPU usage threshold being hit for a period of time. Which means that if you are using a longer window to average the CPU usage, you don't know the distribution of usages, which is what matters.

An example: the watchdog samples every 3 seconds and reacts when the total CPU usage is > 10% for 12 seconds **sustained**.
This means that every 3 seconds everytime the CPU time over those 3 seconds is over 10% total usage, then it will decide that for 3 seconds the usage was over the limit. If it finds that situation for other 3 times, than it triggers.

Now take the suggested implementation which averages over 1 minute, even with sampling every 3 seconds.
Now think of a situation where the sample alternatingly comes up with 100% total CPU usage, and then 0.
Over 1 minute that will be averaged out at 50%, and even though the % is higher than what the watchdog in theory should react to, it hasn't killed osquery.

Instead lets think of a situation where the sample is always at 50%; a part from the fact that it will not arrive to a 1 minute worth of samples, because the watchdog will react before that, but even if did, it would again show 50%.
So in that case there would be no way to differentiate the two, for the purpose of understanding what happened with the watchdog.

Ideally these kinds of metrics should probably be sent as a stream, in part to prevent the issue of not getting all the samples if not quering often enough, and the inherent imprecision if you do query often enough. In part to get all the raw samples and apply all the needed logics at the other end.
Also ideally the lowest sampling interval possible is the best to then extrapolate the CPU usage over bigger intervals in the backend, but obviously there's the problem of the amount of data and the cost of sampling so often.